Adding np.vecdot overload

oleksii-kulyk opened the pull request:
Hello, I am submitting this draft PR because I need help understanding the error message.
I took on the task of trying to implement the overload for numpy.vecdot inside numba.

Starting at [Numpy Supported Features](https://numba.readthedocs.io/en/stable/reference/numpysupported.html#), I figured out that Numba currently has ufuncs to support Numpy behavior for sum and multiplication.
I also found that the following formula in Numpy behaves identically to the numpy.vecdot function:
`sum( moveaxis( arr1, axis1, -1).conjugate() * moveaxis( arr2, axis2, -1), axis = -1 )`

This inspired the first version of my attempted implementation:
```python
def np_vecdot_impl(x1, x2, axis=-1, axes=None):
    if axes is None:
        axis1, axis2 = axis, axis
    else:
        axis1, axis2 = axes[0], axes[1]
    #normalize axis1 and axis2
    left = np.moveaxis(x1, axis1, -1).conjugate()
    right = np.moveaxis(x2, axis2, -1)
    product = np.multiply(left, right)
    result = np.sum(product, axis=-1)
    return result
```

Unfortunately, Numba's moveaxis implementation returns a slicing error with the following traceback:

```
  File "D:\Code Projects\Open Source\numba\numba\np\linalg.py", line 912, in np_vecdot_axis
    return np_vecdot_impl(x1, x2, axis=axis)
  File "D:\Code Projects\Open Source\numba\numba\np\linalg.py", line 1029, in np_vecdot_impl
    left = np.moveaxis(x1, axis1, -1).conjugate()
  File "C:\Users\Oleksiy\miniconda3\envs\numbaenv\lib\site-packages\numpy\_core\numeric.py", line 1503, in moveaxis
    source = normalize_axis_tuple(source, a.ndim, 'source')
  File "C:\Users\Oleksiy\miniconda3\envs\numbaenv\lib\site-packages\numpy\_core\numeric.py", line 1435, in normalize_axis_tuple
    axis = tuple([normalize_axis_index(ax, ndim, argname) for ax in axis])
  File "C:\Users\Oleksiy\miniconda3\envs\numbaenv\lib\site-packages\numpy\_core\numeric.py", line 1435, in <listcomp>        
    axis = tuple([normalize_axis_index(ax, ndim, argname) for ax in axis])
  File "D:\Code Projects\Open Source\numba\numba\core\types\abstract.py", line 189, in __getitem__
    ndim, layout = self._determine_array_spec(args)
  File "D:\Code Projects\Open Source\numba\numba\core\types\abstract.py", line 214, in _determine_array_spec
    raise KeyError(f"Can only index numba types with slices with no start or stop, got {args}.")
KeyError: 'Can only index numba types with slices with no start or stop, got 0.'
```
Seeing that there was an issue with the `moveaxis()` function, I decided to replace is with the `transpose( permutation )` function adding logic for creating the required `permutation`.

This led to my second attempt at implementation, which is in this Draft PR:
```python
def np_vecdot_impl(x1, x2, axis=-1, axes=None):
    if axes is None:
        axis1, axis2 = axis, axis
    else:
        axis1, axis2 = axes[0], axes[1]

    # does the same as np.moveaxis(x1, axis1, -1)
    # but moveaxis doesn't behave well in nopython mode
    perm1 = np.empty(x1.ndim, dtype=np.intp)
    for i in range(x1.ndim):
        if i < axis1:
            perm1[i] = i
        elif i == axis1:
            perm1[i] = x1.ndim - 1
        else:
            perm1[i] = i - 1

    # does the same as np.moveaxis(x2, axis2, -1)
    # but moveaxis doesn't behave well in nopython mode
    perm2 = np.empty(x2.ndim, dtype=np.intp)
    for i in range(x2.ndim):
        if i < axis2:
            perm2[i] = i
        elif i == axis2:
            perm2[i] = x2.ndim - 1
        else:
            perm2[i] = i - 1

    left = np.transpose(x1, perm1).conjugate()
    right = np.transpose(x2, perm2)
    product = np.multiply(left, right)
    result = np.sum(product, axis=-1)
    return result
```

Unfortunately, I ran into a mismatch between the Numba and Python types with the following traceback:
```
  File "D:\Code Projects\Open Source\numba\numba\core\typing\templates.py", line 621, in generic
    disp, new_args = self._get_impl(args, kws)
  File "D:\Code Projects\Open Source\numba\numba\core\typing\templates.py", line 720, in _get_impl
    impl, args = self._build_impl(cache_key, args, kws)
  File "D:\Code Projects\Open Source\numba\numba\core\typing\templates.py", line 793, in _build_impl
    ovf_result = self._overload_func(*args, **kws)
  File "D:\Code Projects\Open Source\numba\numba\np\linalg.py", line 912, in np_vecdot_axis
    return np_vecdot_impl(x1, x2, axis=axis)
  File "D:\Code Projects\Open Source\numba\numba\np\linalg.py", line 1006, in np_vecdot_impl
    if i < axis1:
TypeError: '<' not supported between instances of 'int' and 'Integer'
```
Things I've tried to convert axis1 to int:
- `int(axis1)` 
- `operator.index(axis1)`
- `literally(axis1)
- `axis1 = int(getattr(axis1, "item", lambda: axis1)())`
- `    if hasattr(x, "item"):
        return int(x.item())`
-`    if hasattr(x, "value"):
        return int(x.value)`

I would welcome any advice so that I can carry this PR across the finish line. I have already spent a lot of time on it, so it would be a shame if it was all for nothing.

Thank you,
Alex

oleksii-kulyk committed:
Adding np.vecdot support

sklam commented:
@oleksii-kulyk, The `@overload` has unusual semantic comparing to regular python code. There's a "typing layer" and a "execution layer". The reason you are seeing the type mismatch is because you are implementing the execution logic in the "typing layer".  Please see https://numba.readthedocs.io/en/stable/extending/high-level.html for details.

oleksii-kulyk committed:
Fixed Numba typing issue "...between instances of
'int' and 'Integer'". Ran into a new problem of
"numpy.exceptions.AxisError: source:
axis 1 is out of bounds for array of dimension 0"

oleksii-kulyk committed:
Merge branch 'numba:main' into adding-Numpy-vecdot-support-#10091

sklam commented on numba/np/linalg.py:
```diff
@@ -903,6 +904,142 @@ def impl(a):
+        return impl
+    else:
+        print(x1, x2, axis)
+        return np_vecdot_impl(x1, x2, axis=axis.literal_value)
```
This should be in a `def impl(x1, x2, axis): ...`

According to your traceback, it's calling into actual `numpy` instead of numba compiled code.

Note: the most confusing concept with `@overload` is the distinction between typing and lowering. `@overload` should return a function object that will be compiled. Otherwise (like this call), is triggering with type object instead of actual values.

sklam commented on numba/np/linalg.py:
```diff
@@ -903,6 +904,142 @@ def impl(a):
+    # does the same as np.moveaxis(x1, axis1, -1)
+    # but moveaxis doesn't behave well in nopython mode
+    # Create permutation tuple for x1: move axis1 to the last position
+    print(x1.ndim, axis1)
+    print(x1, x2)
+    perm1 = [i for i in range(x1.ndim) if i != axis1]
+    perm1.append(axis1)
+    
+    # does the same as np.moveaxis(x2, axis2, -1)
+    # but moveaxis doesn't behave well in nopython mode
+    # Create permutation tuple for x2: move axis2 to the last position
+    perm2 = [i for i in range(x2.ndim) if i != axis2]
+    perm2.append(axis2)
+    
+    perm1_tuple = tuple(perm1)
+    perm2_tuple = tuple(perm2)
+    
+    #left = np.transpose(x1, perm1_tuple).conjugate()
+    #right = np.transpose(x2, perm2_tuple)
+    left = np.moveaxis(x1, axis1, -1).conjugate()
+    right = np.moveaxis(x2, axis2, -1)
+    product = np.multiply(left, right)
+    result = np.sum(product, axis=-1)
```
This should be in a `impl` function that is returned by `@overload`

sklam commented on numba/np/linalg.py:
```diff
@@ -903,6 +904,142 @@ def impl(a):
+    array([ 3.,  8., 10.])
+    """
+    # Scalars are not allowed per numpy's implementation
+    if np_support.type_is_scalar(x1) or np_support.type_is_scalar(x2):
+        raise TypingError("np.vecdot() only supported on array types")
```
This should be part of the "typing" in the logic of `@overload`
